Initial feature gate documentation

irbekrm opened the pull request:
This PR:

- attempts to define a loose feature policy just so that we can graduate some of the existing alpha features to beta and communicate to users what they should expect from alpha/beta features

- attempts to document the current mechanism how we implement feature gated API fields (for implementors)

All 'policy' bits are open to discussion, I've largely just pulled things from kube docs and what we discussed in the previous dev meeting

Fixes #988 

Signed-off-by: irbekrm <irbekrm@gmail.com>

munnerz commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+
+### Potential issues
+
+- A user has to remember to set two feature gates, one of the webhook one on the controller for the feature to function. Forgetting to set one of them might result in unexpected behavior
```
I think it's worth noting that it's actually the *operator* or administrator, not the end-user that has to do this (depending on how we define these different roles/personas! 😄)

munnerz commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+
+- A user has to remember to set two feature gates, one of the webhook one on the controller for the feature to function. Forgetting to set one of them might result in unexpected behavior
+
+- A user must remember to remove the alpha fields from their manifests when disabling a previously enabled API feature. Failing to do so might result in unexpected behavior- for example forgetting to remove an feature gated field from a `Certificate` resource might result in failed renewals at some later point when cert-manager's controller will attempt to update the `Certificate` spec, but the w ...
```
👍 your idea of only gating on CREATE and not UPDATE sounds reasonable (though I need to consider/mull over any side-effects of this..). That said, this _is_ why we advise people to not use alpha features and is exactly why they are gated. We could remove the feature *altogether* in release N+1 as well.

irbekrm replied:
Thanks I will also look into how CREATE vs UPDATE would look like

Agree that alpha could be less stable.. but I see a bunch of proposals/issues/PRs that would result in new alpha fields against API, some of which may be more popular than others.. so I think in reality we may end up with this mechanism being used quite a bit

irbekrm replied:
This can be looked into separately and probably won't happen now when a bunch of things have already been implemented using the 2 flags approach

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+
+## Feature gated API fields
+
+Feature gated API fields are implemented using `--feature-gates` flags of cert-manager [webhook](../cli/webhook.md) and [controller](../cli/controller.md).
```
Should they always have the same name? If so, it'd be good to spell that out.

irbekrm replied:
This is really up to the implementor

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+
+Feature gated API fields are implemented using `--feature-gates` flags of cert-manager [webhook](../cli/webhook.md) and [controller](../cli/controller.md).
+
+A feature gated API field is always visible to the user (i.e when running `kubectl explain <some-resource>`), but is only functional if the relevant feature is explicitly enabled via both feature flags.
```
```suggestion
A feature gated API field is always visible to the user (i.e when running `kubectl explain <some-resource>`), but is only functional if the relevant feature is explicitly enabled via feature flags for both the webhook and controller.
```

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+
+A feature gated API field is always visible to the user (i.e when running `kubectl explain <some-resource>`), but is only functional if the relevant feature is explicitly enabled via both feature flags.
+
+If a user attempts to apply a resource with the feature gated field set to a non-nil value, but the feature gate is not enabled, the resource will get rejected by the webhook validation.
```
```suggestion
If a user attempts to apply a resource with the feature gated field set to a non-nil value, but the feature gate is not enabled, the resource will get rejected by webhook validation.
```

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+A feature gated API field is always visible to the user (i.e when running `kubectl explain <some-resource>`), but is only functional if the relevant feature is explicitly enabled via both feature flags.
+
+If a user attempts to apply a resource with the feature gated field set to a non-nil value, but the feature gate is not enabled, the resource will get rejected by the webhook validation.
+This mechanism differs from [the one that Kubernetes uses for feature gated API field implementation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) where the field will be simply set to nil if the feature gate is disabled. We chose to use webhook validation instead to make debugging easier for users who ar ...
```
```suggestion
This mechanism differs from [the one that Kubernetes uses for feature gated API field implementation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) where the field will be simply set to nil if the feature gate is disabled. We chose to use webhook validation instead to make debugging easier for users who are attempting to use the feature gated field, but have forgotten to enable the feature gate.
```

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+
+### Implementation
+
+- Implement the new field and document that it is feature gated and in order to use it feature gates need enabling
```
```suggestion
- Implement the new field and document that it is feature gated and in order to use it the webhook and controller feature gates that need to be enabled
```

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+- Add a new [webhook feature gate](https://github.com/cert-manager/cert-manager/blob/3a055cc2f56c1c2874807af4a8f84d0a1c46ccb4/internal/webhook/feature/features.go#L25-L39) for the field
+- Update webhook validation checks for the relevant resource kind to ensure that if the feature gated field is set, but the webhook feature gate is not enabled, the resource gets rejected
+- Add a new [controller feature gate](https://github.com/cert-manager/cert-manager/blob/2417132b3cd017b5f0974006e03c2b8a540efe3f/internal/controller/feature/features.go#L26-L54) for the field
+- Ensure that any control loops that use the feature, check that the feature gate is actually enabled. (This is required to cover edge cases such if webhook runs a version of cert-manager where the feature is in GA whereas controller runs an older version where the feature is still in experimental state)
```
```suggestion
- Ensure that any control loops that use the feature, check that the feature gate is actually enabled. (This is required to cover edge cases such as if the webhook runs a version of cert-manager where the feature is in GA whereas controller runs an older version where the feature is still in experimental state.)
```

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,48 @@
+
+- A user has to remember to set two feature gates, one of the webhook one on the controller for the feature to function. Forgetting to set one of them might result in unexpected behavior
+
+- A user must remember to remove the alpha fields from their manifests when disabling a previously enabled API feature. Failing to do so might result in unexpected behavior- for example forgetting to remove an feature gated field from a `Certificate` resource might result in failed renewals at some later point when cert-manager's controller will attempt to update the `Certificate` spec, but the w ...
```
```suggestion
- A user must remember to remove the alpha fields from their manifests when disabling a previously enabled API feature. Failing to do so might result in unexpected behavior- for example forgetting to remove a feature gated field from a `Certificate` resource might result in failed renewals at some later point when cert-manager's controller will attempt to update the `Certificate` spec, but the webhook will reject the update due to the feature gated field being set
```

irbekrm commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+The graduation criteria can be different for each feature.
+Generally, we find user feedback most valuable when determining if a feature is sufficiently mature to graduate. If you are using an alpha or beta feature and would like to see it graduate, it would be great if you could give us some feedback about how you use it and whether you find the API useful to initiate graduation process. Feel free to [open a Github issue](https://github.com/cert-manager/ ...
+
+## List of current feature gates
```
Inspiration from https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates

irbekrm commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+
+A feature gate can be toggled on/off using `--feature-gates` flags on cert-manager controller. For feature gated functonality that comes with new API fields there is also a corresponding feature gate on webhook that also needs to be enabled using a `--feature-gates` flag if you want to use it.
+
+**Alpha**
+
+All alpha features are off by default.
+We keep the right to remove or change alpha features without a warning.
+An API field that is part of an alpha feature and requires a webhook feature flag to be used also can be removed from the API without a warning.
+An alpha feature might not work for all versions of Kubernetes supported by cert-manager.
+
+**Beta**
+
+All beta features are off by default.
+Beta features will not be removed, but can be changed.
+If the feature gets changed in incompatible ways, we will provide migration instructions.
+A beta feature will work with all versions of Kubernetes supported by cert-manager.
+
+**GA**
+
+A feature that is GA is on by default and cannot be disabled.
+With regards to API fields and their functionality, we keep Kubernetes API compatibility promise https://cert-manager.io/docs/installation/api-compatibility/.
+The feature flag for a GA feature will eventually be removed.
+
```
This is largely adapted from https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages

irbekrm commented:
Kube also has feature metrics apparently https://kubernetes.io/blog/2023/02/03/sig-instrumentation-spotlight-2023/#:~:text=We%E2%80%99ve%20also%20added%20feature%20metrics%20so%20that%20you%20can%20look%20at%20your%20metrics%20to%20determine%20which%20features%20are%20enabled%20in%20your%20cluster%20at%20a%20given%20time

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,47 @@
+A feature gated API field is always visible to the user (i.e when running `kubectl explain <some-resource>`), but is only functional if the relevant feature is explicitly enabled via feature flags for both the webhook and controller.
+
+If a user attempts to apply a resource with the feature gated field set to a non-nil value, but the feature gate is not enabled, the resource will get rejected by the webhook validation.
+This mechanism differs from [the one that Kubernetes uses for feature gated API field implementation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) where the field will be simply set to nil if the feature gate is disabled. We chose to use webhook validation instead to make debugging easier for users who ar ...
```
If an object is created using a feature gated thing while everything is enabled and then the controller/webhook are disabled, what happens when the controller restarts? Does anything flag for this?

irbekrm replied:
You would get errors in controller logs if there are resources with fields that are no longer valid. I've added a note to warn folks that resources need updating before disabling a previously enabled feature gate flags.

jsoref commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+**Alpha**
+
+All alpha features are off by default.
+We keep the right to remove or change alpha features without a warning.
```
```suggestion
We retain the right to change or remove alpha features without warning.
```

jsoref commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+
+All alpha features are off by default.
+We keep the right to remove or change alpha features without a warning.
+An API field that is part of an alpha feature and requires a webhook feature flag to be used also can be removed from the API without a warning.
```
```suggestion
An API field that is part of an alpha feature and requires a webhook feature flag to be used can also be removed from the API without warning.
```

jsoref commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+All alpha features are off by default.
+We keep the right to remove or change alpha features without a warning.
+An API field that is part of an alpha feature and requires a webhook feature flag to be used also can be removed from the API without a warning.
+An alpha feature might not work for all versions of Kubernetes supported by cert-manager.
```
```suggestion
An alpha feature might not work for all of cert-manager's supported Kubernetes versions.
```

jsoref commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+**Beta**
+
+All beta features are off by default.
+Beta features will not be removed, but can be changed.
```
```suggestion
Beta features will not be removed, but may be changed.
```

jsoref commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+All beta features are off by default.
+Beta features will not be removed, but can be changed.
+If the feature gets changed in incompatible ways, we will provide migration instructions.
+A beta feature will work with all versions of Kubernetes supported by cert-manager.
```
```suggestion
A beta feature will work with all versions of cert-manager's supported Kubernetes versions.
```

jsoref commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+**GA**
+
+A feature that is GA is on by default and cannot be disabled.
+With regards to API fields and their functionality, we keep Kubernetes API compatibility promise https://cert-manager.io/docs/installation/api-compatibility/.
```
Why is this a bare url?

SgtCoDFish commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+---
+title: Features
+description: cert-manager feature policy
+---
```
nitpick: I don't feel like I've got a great suggestion here, but it feels like "features" doesn't quite match what this page describes.  Don't get me wrong - i love the idea of adding this page.

Maybe something like "Unstable Features", "Feature Flags" or something? I think as a user if I see "features" in the sidebar I might expect like a vendor pitch of what the product can do, and I'd skip it. If I saw something like "feature flags" I'd think "oh, more things to try!" and check it out.

What do you think?

jsoref replied:
`experimental`, `evolving`, `under development`

SgtCoDFish commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+
+New cert-manager features and functionality is often initially implemented behind a feature gate. This is so as to not break users with functionality that has not yet been tested in production as well as to give us a chance to remove or modify API fields and functionality following user feedback.
+
+We have alpha and beta features. We do not aim to keep any of the features in alpha or beta stage indefinitely. Feature gating should only be used for functionality that can eventually be turned on by default for all users (GA).
```
question: I think the way in which some features currently behind gates are implemented might break this?

e.g. ValidateCAA is feature gated, but it's [also](https://github.com/cert-manager/cert-manager/blob/b0caedffd915eab1a52e06473dc2b5b527b84460/pkg/controller/acmechallenges/sync.go#L149) a toggle on whether or not to do the validation. We might want to graduate ValidateCAA to GA (or not, it's just an example) but it would seem reasonable for users to want to disable it.

In fact, for this specific feature I'd imagine users would want to opt-in rather than opt-out (and that might even be required so that we don't break people).

Considering this:

> Feature gating should only be used for functionality that can eventually be turned on by default for all users (GA).

For this to be true, we'd have to add a new flag "--validate-caa" which is required _as well as the feature gate_, right? (Or some other way of toggling the feature on in addition to the feature gate)

irbekrm replied:
> For this to be true, we'd have to add a new flag "--validate-caa" which is required as well as the feature gate, right? (Or some other way of toggling the feature on in addition to the feature gate)

Yes, I think that the mechanism for marking a piece of functionality as alpha/beta should be separate from ways how something that should always be opt-in gets toggled on/off. I think that's better than mixing the two up because it simplifies the feature flag mechanism that is already complicated, is how it's implemented in kube and would appear more consistent to users once it's in GA.

It has the downside that when some functionality X that should always be opt-in is in alpha, using it might look like
```
go run <cert-manager-controller> \
 --features-gates=<optional-functionality-x>=true \
 --<enable-optional-functionality-x>
```
but I think this inconvenience is acceptable so that in GA we can have

```
go run <cert-manager-controller> 
 --<enable-optional-functionality-a>
 --<enable-optional-functionality-b>
 --<enable-optional-functionality-x>
```
and not
```
go run <cert-manager-controller> 
 --<enable-optional-functionality-a>
 --<enable-optional-functionality-b>
 --feature-gates=<enable-optional-functionality-x> // X is GA same as A and B, but gets toggled on via --feature-gates flag because it once was an alpha feature
```

I think it's ok if enabling an alpha feature is a bit clunky, perhaps it'd even encourage folks to nudge us to move it through to graduation.

Let me know what you think.

irbekrm replied:
> But since the features are feature gated

😆 

yeah, I think in some cases it might not be possible to know whether something will need to opt-in. For example with Gateway, we originally assumed that those CRDs will at some point become part of core Kubernetes.

SgtCoDFish commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+
+A feature that is GA is on by default and cannot be disabled.
+With regards to API fields and their functionality, we keep Kubernetes API compatibility promise https://cert-manager.io/docs/installation/api-compatibility/.
+The feature flag for a GA feature will eventually be removed.
```
suggestion: I'm not sure if there'd be any reason to ever _actually_ remove one - it feels like a needless way to break people.

It doesn't seem like a huge burden to move old feature flags to `OldFeatureFlags []string`, where if any of them are passed we just log a warning and then move on.

Removing would make sense if we were ever gonna re-use a flag name but I think that could get confusing tbh

irbekrm replied:
Yeah I agree with you, I think removing would make sense if we ended up with a huge number of features gates, so that the help output would become unreadable, but this is currently not the case.

irbekrm replied:
The downside would be that if it's a feature gated opt-in functionality that gets enabled via flag in GA, it could be confusing that it appears in two places. But since this would be non-functional, we could change it if folks complained.

SgtCoDFish commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+## Graduation
+
+The graduation criteria can be different for each feature.
+Generally, we find user feedback most valuable when determining if a feature is sufficiently mature to graduate. If you are using an alpha or beta feature and would like to see it graduate, it would be great if you could give us some feedback about how you use it and whether you find the API useful to initiate graduation process. Feel free to [open a GitHub issue](https://github.com/cert-manager/ ...
```
nitpick: the meetings link is to the website, so it might be worth following the [linking to other pages](https://github.com/cert-manager/website/#task-linking-to-other-documentation-pages) guide.

irbekrm replied:
Thank for the reminder, that guide is great btw 👍🏼 I have changed all internal links.

SgtCoDFish commented on content/docs/installation/features.md:
```diff
@@ -0,0 +1,62 @@
+
+See `--feature-gates` flags on cert-manager controller and webhook to enable any of these features.
+
+- `AdditionalCertificateOutputFormats`. Added in cert-manager 1.7.0. Allows to specify additional formats in which cert-manager will store issued certificates and keys. See [release note](https://cert-manager.io/docs/release-notes/release-notes-1.7#additional-certificate-output-formats). Requires the feature to be enabled on both cert-manager controller and webhook
```
nitpick: another website link here; there are a few more in this section.

SgtCoDFish reviewed (commented):
I really like the overall gist of this. Got a few comments on it overall, but I like the core of this! Thank you for your work in pushing this forwards, it's super valuable!

irbekrm commented:
Thanks for the review folks! I've now addressed all the comments, PTAL

https://deploy-preview-992--cert-manager-website.netlify.app/docs/installation/featureflags/
https://deploy-preview-992--cert-manager-website.netlify.app/docs/contributing/featuregates/

jsoref commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,47 @@
+
+- The person deploying cert-manager has to remember to set two cert-manager feature gates, one of the webhook one on the controller for the feature to function. Forgetting to set one of them might result in unexpected behavior
+
+- A user must remember to remove the alpha fields from their manifests when disabling a previously enabled API feature. Failing to do so might result in unexpected behavior- for example forgetting to remove feature gated field from a `Certificate` resource might result in failed renewals at some later point when cert-manager's controller will attempt to update the `Certificate` spec, but the webh ...
```
Add `.` at end

jsoref commented on content/docs/installation/featureflags.md:
```diff
@@ -0,0 +1,72 @@
+
+- `UseCertificateRequestBasicConstraints`. Added in cert-manager 1.11.0. Makes cert-manager add a basic constraints section to certificate signing requests with the CA constraint set to the correct value. See [`cert-manager#5552`](https://github.com/cert-manager/cert-manager/pull/5552)
+
+- `ValidateCAA`. Added in cert-manager 0.7.2. CAA checking when issuing a certificate.
```
Can you add a see link for this one?

irbekrm replied:
There's no description/documentation for this as far as I am aware, it was implemented pre-v1 where we didn't have more descriptive release notes yet

SgtCoDFish commented on content/docs/installation/featureflags.md:
```diff
@@ -0,0 +1,72 @@
+description: Using feature gated functionality
+---
+
+New cert-manager features and functionality is often initially implemented behind a feature gate. This is so as to not break users with functionality that has not yet been tested in production as well as to give us a chance to remove or modify API fields and functionality following user feedback.
```
nitpick: grammar
```suggestion
New cert-manager features and functionality are often initially implemented behind a feature gate. This is so as to not break users with functionality that has not yet been tested in production as well as to give us a chance to remove or modify API fields and functionality following user feedback.
```

SgtCoDFish commented on content/docs/installation/featureflags.md:
```diff
@@ -0,0 +1,72 @@
+API without warning. An alpha feature might not work for all cert-manager's
+supported Kubernetes versions. If you want to disable a previously enabled alpha
+feature gate, you should make sure that you have updated any resources that have API
+fields set that are only valid if the feature gate is enabled, else the resources
+will end up in invalid state.
```
nitpick: missing "an"

```suggestion
fields set that are only valid if the feature gate is enabled, or else the resources
will end up in an invalid state.
```

SgtCoDFish commented on content/docs/manifest.json:
```diff
@@ -472,6 +476,10 @@
               "path": "/docs/contributing/kind.md"
             },
+            {
+              "title": "Implementing feature gates",
```
nitpick: The rest of the titles in this section use title case so this one stands out a little

```suggestion
              "title": "Implementing Feature Gates",
```

SgtCoDFish commented on content/docs/contributing/featuregates.md:
```diff
@@ -0,0 +1,47 @@
+
+- Kubernetes [API change design](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+
+[^1]: for example, functionality that might change in the future in response to user feedback
```
serious nitpick (non-blocking): Does this look better if capitalized? Really, really doesn't matter though!

```suggestion
[^1]: For example, functionality that might change in the future in response to user feedback
```

SgtCoDFish reviewed (commented):
Spotted a few typos and bits but I'm happy to merge this now!

SgtCoDFish reviewed (approved):
/lgtm
/approve

Thanks for this! I think it's good to go - it can always evolve more later if we end up needing that!

I guess this does introduce a need to rework some of the feature gated stuff but I don't really see a huge rush!
